### PR TITLE
feat: Add client-side logging support

### LINKED ISSUE #42: Implement logging in client
Issue to implement the client-side section of https://github.com/strimzi/strimzi-ui/pull/41

### PR DESCRIPTION
**Pull request by ajborley**
- Adds the websocket listener to the server.
- Adds support on the client to send logging events through the
websocket via a new `useLogger` hook.

Contributes to: #42

Signed-off-by: Andrew Borley <borley@uk.ibm.com>

### CODE REVIEW

**Comment by ajborley** [comment]
Draft PR to check if folks think I'm on the right track... still working on how to test this all.

File: client/Hooks/useLogger/Hook.ts
Comment (pmuir) [inline client/Hooks/useLogger/Hook.ts]: Feels like we need this before we can merge?

File: client/Hooks/useLogger/Hook.ts
Comment (nictownsend) [inline client/Hooks/useLogger/Hook.ts]: Could we have the standard `log` `debug` etc as curried methods instead of passing the level in whenever you call the logger method?

```
const genLog = (logLevel) => (message) => {
  ...
 buffer.push({logLevel, message});
  ...
};


return {
  info : genLog("info);
  ...
}
```

File: client/Hooks/useLogger/Hook.ts
Comment (nictownsend) [inline client/Hooks/useLogger/Hook.ts]: UUID? Or something sent up from server when the socket opens?

File: server/log/router.ts
Comment (nictownsend) [inline server/log/router.ts:44]: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/c1c23b2dcaf1a419992379a2b5bf7e14fe829f7a/types/ws/index.d.ts#L118 - any chance we could end up with buffers? Is it client that decides the format?

File: server/log/router.ts
Comment (nictownsend) [inline server/log/router.ts]: https://github.com/pinojs/pino/blob/HEAD/docs/api.md#logging-method-parameters - looks like we don't need second param as our event contains `msg` field

File: server/log/router.ts
Comment (nictownsend) [inline server/log/router.ts:67]: Can we log out a request ID / client ID?

File: server/log/router.ts
Comment (nictownsend) [inline server/log/router.ts]: Pedantic, but can you rename to `messageHandler` (as it's not itself handling it, but the return function.

File: server/log/router.ts
Comment (nictownsend) [inline server/log/router.ts:31]: Is this just belts and braces on the error case? Could return `404` as it's not a valid call for non web?

File: server/main.ts
Comment (nictownsend) [inline server/main.ts]: How does `expressApp` know which endpoint to call? Is it simply that we're listening on `host:port/*` so any websocket request will be upgraded, but then sending messages will fail unless it's on a path that has `isWs` support?

File: server/test_common/commonServerSteps.ts
Comment (nictownsend) [inline server/test_common/commonServerSteps.ts]: `toHaveBeenCalled` and related - https://jestjs.io/docs/en/expect#tohavebeencalled - is a neater API than `mock.calls.x`

**Review by nictownsend — CHANGES_REQUESTED** [review]

File: client/Hooks/useLogger/Hook.ts
Comment (ajborley) [reply client/Hooks/useLogger/Hook.ts]: Good idea - done.

File: client/Hooks/useLogger/Hook.ts
Comment (ajborley) [reply client/Hooks/useLogger/Hook.ts]: Yep - have picked up the `uuid` package.

File: server/log/router.ts
Comment (ajborley) [reply server/log/router.ts:44]: Yep - the `useLogger` hook will always send strings so this should be fine.

File: server/log/router.ts
Comment (ajborley) [reply server/log/router.ts]: Good spot, have removed.

File: server/log/router.ts
Comment (ajborley) [reply server/log/router.ts:67]: The close event only receives the code and reason string: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/ws/index.d.ts#L86 so we would need to encode the client ID in the reason string. I will look into this when I hook into the context, as we may want to explicitly close the logging websocket on client unload.

File: server/log/router.ts
Comment (ajborley) [reply server/log/router.ts]: Done

File: server/log/router.ts
Comment (ajborley) [reply server/log/router.ts:31]: We expect HTTP GET calls to the `/log` endpoint to be websocket upgrade requests. If we get a request that is not an upgrade, then we return the [426](https://httpstatuses.com/426) response, as we don't serve any other http traffic from `/log`.

File: server/main.ts
Comment (ajborley) [reply server/main.ts]: Yep, that's right. The code in main.ts (https://github.com/ajborley/strimzi-ui/blob/42-client-side-logging/server/main.ts#L74-L90) watches for websocket upgrade requests, and adds the `isWs` boolean and `ws` websocket to the req and res respectively. So currently, any endpoint that receives an upgrade request will set up a websocket, but only modules that implement the `websocket.onmessage`, etc handlers will actually do anything.

I expect this is not what we want - we should probably add something to the module configuration that can be used to decide whether the websocket upgrade request is permitted or not. What do you think?

File: server/test_common/commonServerSteps.ts
Comment (ajborley) [reply server/test_common/commonServerSteps.ts]: Forgot all about the `toHaveBeenCalled*` API.  :) Updated

File: client/Hooks/useLogger/Hook.ts
Comment (ajborley) [reply client/Hooks/useLogger/Hook.ts]: Agreed, and the same for the other TODOs (https://github.com/strimzi/strimzi-ui/pull/85/files#diff-3e52b502269a41a2c06ce466fd7d975b726b3557bc98a54d84b74b2bf958f1baR9-R12)

File: server/main.ts
Comment (nictownsend) [reply server/main.ts]: I wonder if we need a break out session on this.

It seems like we're possibly shoehorning in the `log` module into express, when actually this is a different paradigm. My concern is that by appending `ws` to `req` we lose reference to it outside of the initial request.

Without extending express or using old express middleware - it feels like we need the ability to control "websocket" modules too - maybe, by having the concept of attaching multiple endpoints (https://www.npmjs.com/package/ws#multiple-servers-sharing-a-single-https-server) - we then destroy socket if it's not on a path that's acceptable?

We can then have "express" modules and "ws" modules (with a similar `addModule` for ws that takes in a logger)

**Comment by matthew-chirgwin** [comment]
Requesting (re) reviews please @nictownsend / @pmuir !

File: client/Hooks/useLogger/README.md
Comment (nictownsend) [inline client/Hooks/useLogger/README.md]:
```suggestion
### `useLogger(componentName: string)`
```

File: client/Hooks/useLogger/Hook.ts
Comment (nictownsend) [reply client/Hooks/useLogger/Hook.ts]: Is this still blocking merge? Feels like yes?

File: client/Hooks/useLogger/useLogger.steps.tsx
Comment (nictownsend) [inline client/Hooks/useLogger/useLogger.steps.tsx]: May be undefined if the `Given` was never called

File: client/Hooks/useLogger/useLogger.steps.tsx
Comment (nictownsend) [inline client/Hooks/useLogger/useLogger.steps.tsx]: Worry is that if jest runs tests in parallel, or someone else uses this `Given` we'll get port clashes. But if this is fixed via the context change as a later commit (or new PR) I'm less worried. Wonder if when you add in the `context` we can supply a mock ws client via context instead of having to start up real servers in the test.

File: client/Hooks/useLogger/useLogger.steps.tsx
Comment (nictownsend) [inline client/Hooks/useLogger/useLogger.steps.tsx:157]: Eugh, that's unfortunate syntax :(

File: server/main.ts
Comment (nictownsend) [inline server/main.ts:88]: Please see previous comment re breakout - not sure how safe this is for GC purposes that we lose reference to the ws (to be able to close it etc)

**Review by nictownsend — CHANGES_REQUESTED** [review]

File: client/Hooks/useLogger/Hook.ts
Comment (dlabaj) [inline client/Hooks/useLogger/Hook.ts]: Are these all required?

File: client/Hooks/useLogger/Hook.ts
Comment (dlabaj) [inline client/Hooks/useLogger/Hook.ts]: Did we want to declare these types instead of using inference?

**Review by dlabaj — APPROVED** [review]
A few comments otherwise looks good.

File: client/Hooks/useLogger/Hook.ts
Comment (pmuir) [reply client/Hooks/useLogger/Hook.ts]: I think it does. Opening a websocket for every log request defeats the purpose of using a websocket.

**Review by pmuir — CHANGES_REQUESTED** [review]
TODOs need resolving, otherwise LGTM

File: server/main.ts
Comment (matthew-chirgwin) [reply server/main.ts:88]: @ajborley / @nictownsend has this chat happened yet? Would be keen to participate. If the `ws` was placed in `res.locals`, it will be scoped to the life of the request (ie socket lifetime). Express will then discard the `locals`, `ws` included. Thoughts?

**Review by matthew-chirgwin — COMMENTED** [review]
LGTM - re the TODOs, the first stage of config work has now landed - would it be possible to get the various TODOs re config usage to be updated to use that, which I can then build upon in #99 ?

File: client/Hooks/useLogger/README.md
Comment (ajborley) [reply client/Hooks/useLogger/README.md]: Fixed

File: client/Hooks/useLogger/useLogger.steps.tsx
Comment (ajborley) [reply client/Hooks/useLogger/useLogger.steps.tsx]: Check added

File: client/Hooks/useLogger/useLogger.steps.tsx
Comment (ajborley) [reply client/Hooks/useLogger/useLogger.steps.tsx]: The `new Server(..)` line here is a [`mock-socket`](https://www.npmjs.com/package/mock-socket) server, which just hooks mock-clients to mock-servers using the supplied server URL as a key. No real servers are being started here :)

File: client/Hooks/useLogger/useLogger.steps.tsx
Comment (ajborley) [reply client/Hooks/useLogger/useLogger.steps.tsx:157]: Agreed, ugly. Couldn't spot any alternative way to use `And` for the same code as a `Given`, `When` or `Then`. If you have any good ideas... :)

File: server/main.ts
Comment (ajborley) [reply server/main.ts:88]: I think this will be superseded when Matt's websocket handling changes go in, so I've left it as-is for now.

File: client/Hooks/useLogger/Hook.ts
Comment (ajborley) [reply client/Hooks/useLogger/Hook.ts]: These are all the levels available in the `pino` library used on the server. Only those at the `pino` logging level will actually get emitted to the log.

File: client/Hooks/useLogger/Hook.ts
Comment (ajborley) [reply client/Hooks/useLogger/Hook.ts]: This code has been updated, hopefully the types are clearer now.

File: client/Contexts/Logging/Logging.types.ts
Comment (JQrdan) [inline client/Contexts/Logging/Logging.types.ts]: Would this be nicer as an enum?

File: client/Contexts/Logging/README.md
Comment (JQrdan) [inline client/Contexts/Logging/README.md]:
```suggestion
This context will store the logging WebSocket client connection and the buffer of logging messages used in the (useLogger)[../../Hooks/useLogger] hook. This allows the useLogger hook to use a single WebSocket connection for all logging calls.
```

File: client/Hooks/useLogger/Hook.ts
Comment (JQrdan) [inline client/Hooks/useLogger/Hook.ts]: You can streamline typing like this:

```suggestion
const shouldLogComponent = (
  componentName: string
): boolean => {
```

then `shouldLogComponent` will infer the type from the right side. It saves having to write parameters and their types twice, which becomes painful when you change them

File: client/Hooks/useLogger/Hook.ts
Comment (JQrdan) [inline client/Hooks/useLogger/Hook.ts]:
```suggestion
const getWebSocketAddress = (): string => {
```

File: client/Hooks/useLogger/Hook.ts
Comment (JQrdan) [inline client/Hooks/useLogger/Hook.ts]:
```suggestion
export const useLogger = (
  componentName: string
): LoggerType => {
```

File: client/Hooks/useLogger/useLogger.steps.tsx
Comment (JQrdan) [inline client/Hooks/useLogger/useLogger.steps.tsx:52]: Not a review comment but just wanted to say this is amusing 😂

File: client/Hooks/useLogger/useLogger.steps.tsx
Comment (JQrdan) [inline client/Hooks/useLogger/useLogger.steps.tsx:87]: I'm curious why you need this eslint comment? I though renderHook behaved so that it should abide by all the rules

**Review by JQrdan — CHANGES_REQUESTED** [review]

File: client/Hooks/useLogger/Hook.ts
Comment (ajborley) [reply client/Hooks/useLogger/Hook.ts]: Latest commit resolves this via context

File: client/Contexts/Logging/Logging.types.ts
Comment (ajborley) [reply client/Contexts/Logging/Logging.types.ts]: Yep, have updated.

File: client/Contexts/Logging/README.md
Comment (ajborley) [reply client/Contexts/Logging/README.md]: Done

File: client/Hooks/useLogger/Hook.ts
Comment (ajborley) [reply client/Hooks/useLogger/Hook.ts]: Nice! Thanks for that, makes it much tidier :)

File: client/Hooks/useLogger/Hook.ts
Comment (ajborley) [reply client/Hooks/useLogger/Hook.ts]: Done

File: client/Hooks/useLogger/Hook.ts
Comment (ajborley) [reply client/Hooks/useLogger/Hook.ts]: Done

File: client/Hooks/useLogger/useLogger.steps.tsx
Comment (ajborley) [reply client/Hooks/useLogger/useLogger.steps.tsx:52]: I know :)

File: client/Hooks/useLogger/useLogger.steps.tsx
Comment (ajborley) [reply client/Hooks/useLogger/useLogger.steps.tsx:87]: Without it the `renderHook` line fails the linter, even though I'm following the [renderHook examples](https://react-hooks-testing-library.com/usage/basic-hooks) (I'm guessing the lint rules just aren't smart enough yet):
```
/strimzi-ui/client/Hooks/useLogger/useLogger.steps.tsx
  84:41  error  React Hook "useLogger" cannot be called inside a callback. React Hooks must be called in a React function component or a custom React Hook function  (react-hooks/rules-of-hooks)
```

Comment (JQrdan) [reply client/Hooks/useLogger/useLogger.steps.tsx:87]: Very strange, in my code I have `const { result } = renderHook(() => useRouteConfig(input));` and it's happy

**Review by JQrdan — APPROVED** [review]
LGTM

File: server/main.ts
Comment (matthew-chirgwin) [reply server/main.ts:88]: Yep - I will update the approach in my next PR

**Review by matthew-chirgwin — APPROVED** [review]

**Review by scholzj — APPROVED** [review]

**Pull request merged**
